Add support for 32-bit ARC (legacy ARC)

cyrozap opened the pull request:
This PR adds support for the legacy, pre-ARCompact ARC ISA with 32-bit instructions. While similar to ARCompact in its programming model and instruction semantics, the instruction encoding is _completely different_ from that of the ARCompact ISA and its 16/32-bit instructions. Since this legacy instruction set tends to be referred to as just the "ARC" ISA in documentation and binutils/GCC, to avoid confusion with the modern ARC ISAs I've named this one "ARC32" in the language definitions.

The ISA manual for this instruction set can be found [here](http://me.bios.io/images/c/c6/ARC4._Programmers_reference.pdf).

As far as I can tell, this ISA was last used in the ARCtangent-A4 processor, which [was used for the Intel Management Engine in chipsets prior to the release of the Intel 5 Series (Ibex Peak) chipsets](https://recon.cx/2014/slides/Recon%202014%20Skochinsky.pdf#page=16). This ISA is also used in all versions of the [Broadcom Crystal HD](https://en.wikipedia.org/wiki/Broadcom_Crystal_HD) video decoder chip (whose firmware can be found [here](https://github.com/yeradis/crystalhd/tree/0d08f182397d732dec70099a6a0af835b677f8df/firmware/fwbin)). Hopefully this demonstrates that the core popular enough to justify inclusion in Ghidra.

Unfortunately, when it comes to C toolchains, support for this ISA was removed from GCC nearly a decade ago, and I have been unable to build the last version that supported it (4.5.4). Fortunately, there is a pre-built ARC toolchain (available [here](https://www.maintech.de/support/toolchains/)) that includes both binutils and GCC and can compile for both big-endian and little-endian ARC. In addition to the pre-built toolchain, binutils 2.25.1 (the last version to support this ISA) can still easily be built from source (without modification), and can be used to assemble and disassemble code for this architecture.

I've tested this processor module's disassembly of some of the libraries included in the pre-built toolchain, and compared the output to that of `objdump -d` from binutils 2.25.1 for the same libraries, and they seem to match, but it'd be nice to have some more automated testing of this, both to confirm that the disassembly listing is correct and to validate the p-code implementation of the instruction semantics. Unfortunately, I don't really understand how to enable these tests, so I'd appreciate some assistance with that.

This module is still missing support for some instructions (MAX, MIN, NORM, see TODOs) as well as semantics for some instructions (I have no idea how to implement LP or the "only execute the delay slot when jumping" for branch/jump instructions--see FIXMEs), so I'd appreciate some assistance here as well.

I had originally intended the directory for this processor module to contain all the variants of the ARC ISA, including ARC32, ARCompact, and later ARC ISAs. This is why I just named the directory "ARC" and not "ARC32", but that was before I discovered that someone else had already started adding support for ARCompact in PR #3006. At this point, I'm not sure if you'd rather I rename this module to ARC32, have #3006 rebase off of this PR and move their code under "ARC", or keep everything separate like it is now. I expect that fully implementing ARC32 will take some more time, so maybe it'd be best to keep ARC32 and ARCompact separate for now and only combine them later once they've both been merged into Ghidra. Or maybe we should just keep them separate indefinitely since their instruction encodings are so different.

I'm eager for any advice, suggestions, and comments that anyone might offer.

niooss-ledger commented:
Hello,
Thanks for sharing this work! I am the author of PR #3006 (support for ARCompact) and I am glad other people are sharing the work their do on similar ISA.

I do not have any strong opinion on whether to keep `Ghidra/Processors/ARCompact/` separate or to merge it in `Ghidra/Processors/ARC/`. For information, there also exists ARCv2 and ARCv3 ISA. I know that ARCv2 is similar as ARCompact but different enough to be quite complex to integrate cleanly with the ARCompact definition (some parts of the opcode instruction space set were completely modified with other instructions, for example). So there are two possibilities and both make sense (in my humble opinion): either having ARC32, ARCompact, ARCv2 and ARCv3 in a single directory named `ARC`, or having separate directories for each ISA.

About your missing instructions, you might be interesting in my commits https://github.com/NationalSecurityAgency/ghidra/pull/3006/commits/685727c8d57e744f550fc763849e49b8d4075012 (which implements `min` and `max` in ARCompact) and https://github.com/NationalSecurityAgency/ghidra/pull/3006/commits/bf71b0ced1393e877ee94078ad155bc999f97ae3 (which implements `norm` in Java, to be able to run Ghidra pcodetests in a way that I documented in the description of https://github.com/NationalSecurityAgency/ghidra/pull/3006/commits/f4262bc2b5b4efb32d455406f9d25ec5f3a0e7f7).

To implement `LP`, it is actually a case which is documented in Ghidra's official documentation on https://ghidra.re/courses/languages/html/sleigh_context.html and more precisely this note:
> The processor supports instructions that temporarily affect the execution of the immediately following instruction(s). For example, many processors support hardware loop instructions that automatically cause the following instructions to repeat without an explicit instruction causing the branching and loop counting.

In ARCompact support, I implemented `lp` using context variables, and this was actually complex enough that I was able to give a talk about it in a conference some weeks ago, https://www.sstic.org/2021/presentation/analyzing_arcompact_firmware_with_ghidra/ . You may be interested to read the section 4 *Loop instruction* of the article (the article is in English and the talk and slides are in French) to better understand what I did.

Thanks again for your work!

cyrozap commented:
Thank you, @niooss-ledger, that actually helps a lot. And thank you for contributing the ARCompact support--you've probably saved me days (or more) worth of effort that I would have otherwise had to spend working on that myself.

> I do not have any strong opinion on whether to keep `Ghidra/Processors/ARCompact/` separate or to merge it in `Ghidra/Processors/ARC/`. For information, there also exists ARCv2 and ARCv3 ISA. I know that ARCv2 is similar as ARCompact but different enough to be quite complex to integrate cleanly with the ARCompact definition (some parts of the opcode instruction space set were completely modified with other instructions, for example). So there are two possibilities and both make sense (in my humble opinion): either having ARC32, ARCompact, ARCv2 and ARCv3 in a single directory named `ARC`, or having separate directories for each ISA.

Yeah, I feel like this is a similar situation to the "ARM" module since the ARM ISA variants (e.g., Thumb-2, ARMv7, ARMv8) in there all have a lot in common, but also differ from each other in big ways, so I thought it would make the most sense to do the same thing with all the ARC ISA variants. But ultimately, it's up to the project maintainers how they want to keep the processor modules organized, and I don't really mind one way or the other how that's done so long as our PRs get merged.

> About your missing instructions, you might be interesting in my commits [685727c](https://github.com/NationalSecurityAgency/ghidra/commit/685727c8d57e744f550fc763849e49b8d4075012) (which implements `min` and `max` in ARCompact) and [bf71b0c](https://github.com/NationalSecurityAgency/ghidra/commit/bf71b0ced1393e877ee94078ad155bc999f97ae3) (which implements `norm` in Java, to be able to run Ghidra pcodetests in a way that I documented in the description of [f4262bc](https://github.com/NationalSecurityAgency/ghidra/commit/f4262bc2b5b4efb32d455406f9d25ec5f3a0e7f7)).

Thanks, I'll take a look.

> To implement `LP`, it is actually a case which is documented in Ghidra's official documentation on https://ghidra.re/courses/languages/html/sleigh_context.html and more precisely this note:
> 
> > The processor supports instructions that temporarily affect the execution of the immediately following instruction(s). For example, many processors support hardware loop instructions that automatically cause the following instructions to repeat without an explicit instruction causing the branching and loop counting.

I noticed that line in the documentation while I was writing this module, but unfortunately there weren't any examples on how to use context variables to do that, and I wasn't able to figure it out on my own.

> In ARCompact support, I implemented `lp` using context variables, and this was actually complex enough that I was able to give a talk about it in a conference some weeks ago, https://www.sstic.org/2021/presentation/analyzing_arcompact_firmware_with_ghidra/ . You may be interested to read the section 4 _Loop instruction_ of the article (the article is in English and the talk and slides are in French) to better understand what I did.

Thank you, that really helps a lot. I'll try implementing that, too.

> Thanks again for your work!

Thanks again to you as well!

GhidorahRex commented:
Hi @cyrozap , you've clearly put some work into this. Let me respond to a couple of your comments so you can keep moving forward with the development.

> This PR adds support for the legacy, pre-ARCompact ARC ISA with 32-bit instructions. While similar to ARCompact in its programming model and instruction semantics, the instruction encoding is _completely different_ from that of the ARCompact ISA and its 16/32-bit instructions. Since this legacy instruction set tends to be referred to as just the "ARC" ISA in documentation and binutils/GCC, to avoid confusion with the modern ARC ISAs I've named this one "ARC32" in the language definitions.

This seems reasonable. They should probably go in one project, similar to the PIC processors (and *not* similar to the SuperH processor).

> The ISA manual for this instruction set can be found [here](http://me.bios.io/images/c/c6/ARC4._Programmers_reference.pdf). Synopsis seems to have removed any references to this ISA from their website, so unfortunately it doesn't seem to be possible to get this manual from them directly.

We can use whatever the latest version of the manual is. We don't include the actual pdfs due to copyright issues, but we can reference it however we need to in the index file.

> As far as I can tell, this ISA was last used in the ARCtangent-A4 processor, which [was used for the Intel Management Engine in chipsets prior to the release of the Intel 5 Series (Ibex Peak) chipsets](https://recon.cx/2014/slides/Recon%202014%20Skochinsky.pdf#page=16). This ISA is also used in all versions of the [Broadcom Crystal HD](https://en.wikipedia.org/wiki/Broadcom_Crystal_HD) video decoder chip (whose firmware can be found [here](https://github.com/yeradis/crystalhd/tree/0d08f182397d732dec70099a6a0af835b677f8df/firmware/fwbin)). Hopefully this demonstrates that the core popular enough to justify inclusion in Ghidra.
> 
> Unfortunately, when it comes to C toolchains, support for this ISA was removed from GCC nearly a decade ago, and I have been unable to build the last version that supported it (4.5.4). Fortunately, there is a pre-built ARC toolchain (available [here](https://www.maintech.de/support/toolchains/)) that includes both binutils and GCC and can compile for both big-endian and little-endian ARC. In addition to the pre-built toolchain, binutils 2.25.1 (the last version to support this ISA) can still easily be built from source (without modification), and can be used to assemble and disassemble code for this architecture.

We can probably cobble something together with the toolchains to make something we can build the binaries with (I hope). Usually, the biggest issue is support for some of the standard libraries. But with GCC and binutils I'm pretty confident we will get something that works well.

> I've tested this processor module's disassembly of some of the libraries included in the pre-built toolchain, and compared the output to that of `objdump -d` from binutils 2.25.1 for the same libraries, and they seem to match, but it'd be nice to have some more automated testing of this, both to confirm that the disassembly listing is correct and to validate the p-code implementation of the instruction semantics. Unfortunately, I don't really understand how to enable these tests, so I'd appreciate some assistance with that.

agreed.

> This module is still missing support for some instructions (MAX, MIN, and NORM--see TODOs) as well as semantics for some instructions (I have no idea how to implement LP or the "only execute the delay slot when jumping" for branch/jump instructions--see FIXMEs), so I'd appreciate some assistance here as well.

Look at sparcv9.sinc starting at line 681 for the delayslot fix. Once you see it, it's easy to understand. Just use the conditional statements to determine whether to jump to the next instruction or not.

For "only execute the delayslot when jumping" you want `if !(condition) goto inst_next; delayslot(1); goto [jump]`
For "only execute delayslot when not jumping" you want `if (condition) goto [jump]; delayslot(1);`

For the looping, I would take a look at @niooss-ledger's implementation as well. It's pretty good, but I haven't fully reviewed it yet.

> I had originally intended the directory for this processor module to contain all the variants of the ARC ISA, including ARC32, ARCompact, and later ARC ISAs. This is why I just named the directory "ARC" and not "ARC32", but that was before I discovered that someone else had already started adding support for ARCompact in PR #3006. At this point, I'm not sure if you'd rather I rename this module to ARC32, have #3006 rebase off of this PR and move their code under "ARC", or keep everything separate like it is now. I expect that fully implementing ARC32 will take some more time, so maybe it'd be best to keep ARC32 and ARCompact separate for now and only combine them later once they've both been merged into Ghidra. Or maybe we should just keep them separate indefinitely since their instruction encodings are so different.
> 
> I'm eager for any advice, suggestions, and comments that anyone might offer.

Good luck!

cyrozap commented:
I just realized that I missed a bunch of instruction encodings (e.g., `op, #shimm, #shimm`, where instruction fields `B` and `C` are both set to either `r61` or `r63`), auxiliary registers are still not fully implemented, and the way I've implemented setting condition flags may not be correct (and can be optimized somewhat), so I've converted this PR into a draft to avoid it being merged before all of that has been fixed.

Apologies for the false start--I'll be sure to mark this PR as ready only once I'm sure that the decoding and decompilation are working properly.

To make it easier to keep track (and so I don't forget), here's the current status of the various issues with this processor module:

- [x] Properly handle delay slots for conditional branch/jump instructions. 1550881
- [x] Implement LP instruction. c7890ff
- [ ] Implement MAX instruction.
- [ ] Implement MIN instruction.
- [ ] Implement NORM instruction.
- [ ] Update DWARF register mappings based on information from GCC.
- [ ] Update compiler spec based on information from GCC.
- [ ] Add p-code tests.
- [ ] Implement `op, #shimm, #shimm` encoding for all the instructions that can use it.
- [ ] Modify auxiliary register implementation to use a memory space instead of register space.
- [ ] Fix condition flag updates for arithmetic operations.

tmc1245 commented:
- [] @tmc1245

tmc1245 commented:
```

```
